pkg/kubelet/kuberuntime/sync_action.go

```diff
@@ -0,0 +1,247 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kuberuntime
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api"
+	runtimeapi "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	proberesults "k8s.io/kubernetes/pkg/kubelet/prober/results"
+	"k8s.io/kubernetes/pkg/util/flowcontrol"
+)
+
+// statusGroup groups the sandbox statuses, init container statuses and app container statues.
+type statusGroup struct {
+	sandboxStatuses       []*runtimeapi.PodSandboxStatus
+	initContainerStatuses []*kubecontainer.ContainerStatus
+	appContainerStatuses  []*kubecontainer.ContainerStatus
+}
+
+// newStatusGroup returns a statusGroup object that contains the sandbox statuses,
+// init container statuses and the app container statuses.
+func newStatusGroup(pod *api.Pod, podStatus *kubecontainer.PodStatus) *statusGroup {
+	var initContainerStatuses []*kubecontainer.ContainerStatus
+	var appContainerStatuses []*kubecontainer.ContainerStatus
+
+	initContainers := make(map[string]struct{})
+	appContainers := make(map[string]struct{})
+
+	for _, c := range pod.Spec.InitContainers {
+		initContainers[c.Name] = struct{}{}
+	}
+	for _, c := range pod.Spec.Containers {
+		appContainers[c.Name] = struct{}{}
+	}
+
+	for _, s := range podStatus.ContainerStatuses {
+		if _, ok := initContainers[s.Name]; ok {
+			initContainerStatuses = append(initContainerStatuses, s)
+		}
+		if _, ok := appContainers[s.Name]; ok {
+			appContainerStatuses = append(appContainerStatuses, s)
+		}
+	}
+
+	return &statusGroup{
+		sandboxStatuses:       podStatus.SandboxStatuses,
+		initContainerStatuses: initContainerStatuses,
+		appContainerStatuses:  appContainerStatuses,
+	}
+}
+
+// containerKillInfo describes a running container that needs to be killed, and the reason for that.
+type containerKillInfo struct {
+	// The status of the container, it contains the
+	// container ID, name, restart count, finished timestamp.
+	// The status will be passed to containerStartInfo if the container
+	// needs to be restart.
+	status *kubecontainer.ContainerStatus
+	// The reason why the container will be killed, e.g. due to failures.
+	reason string
+	// The boolean indicates whether the container will be restarted.
+	restart bool
```

Comment:
This is just because we don't add these to-be-restart containers into the list when computing what containers to kill. Instead we just mark them with a boolean, so later when computing what containers to start, we look at these to-be-killed containers, and add them to the to-be-start list if they contain the boolean.

See https://github.com/yifan-gu/kubernetes/blob/f27520a6f9402da381bd36b871ad67487aa31143/pkg/kubelet/kuberuntime/sync_action.go#L676-L685